### PR TITLE
All Resources Unit Testing Set Up + Example

### PR DESCRIPTION
**Pull request by BriannaRoskind**
Note that this PR is meant to be merged on top of PRs [#66](https://github.com/aws-controllers-k8s/sagemaker-controller/pull/66) and [#68](https://github.com/aws-controllers-k8s/sagemaker-controller/pull/68)

Summary: Added example Create with invalid input unit test for all resources.

Code Coverage: 26.0%

New Files Per Resource [resource-name]:
(4 files/feature * 14 features = 56 new files)

- pkg/resource/**[resource-name]**/manager_test_suite_test.go
   - Contains resource specific helper functions for unit testing 
- pkg/resource/**[resource-name]**/testdata/test_suite.yaml
   - Contains the written out unit test
- pkg/resource/**[resource-name]**/testdata/endpoint/v1alpha1/**[resource-abbreviation]**_cmd_invalid_before_create.yaml
   - Example input / desired state file that is used for testing Create functionality with an InvalidParameterValue.
- pkg/resource/**[resource-name]**/testdata/endpoint/v1alpha1/**[resource-abbreviation]**_cmd_invalid_create_attempted.yaml
   - Example output / expected state file that is used for testing Create functionality with an InvalidParameterValue.

### CODE REVIEW

**Review by akartsky — APPROVED** [review]

**Comment by akartsky** [comment]
Can you add current coverage info to the PR

File: pkg/resource/endpoint_config/testdata/endpoint_config/cr/ec_cmd_invalid_create_attempted.yaml [hunk 1]
```diff
@@ -0,0 +1,15 @@
+apiVersion: sagemaker.services.k8s.aws/v1alpha1
+kind: EndpointConfig
+metadata:
+  creationTimestamp: null
+  name: unit-testing-endpoint-config
+spec:
+  endpointConfigName: null
```
Comment (surajkota) [inline pkg/resource/endpoint_config/testdata/endpoint_config/cr/ec_cmd_invalid_create_attempted.yaml:7]: latest will not have spec fields as null?

File: pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_before_create.yaml [hunk 2]
```diff
@@ -0,0 +1,14 @@
+apiVersion: sagemaker.services.k8s.aws/v1alpha1
+kind: Endpoint
+metadata:
+kk  name: unit_testing_endpoint
```
Comment (surajkota) [inline pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_before_create.yaml:4]: typo: kk

are these not getting caught in tests?

File: pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_create_attempted.yaml [hunk 3]
```diff
@@ -0,0 +1,21 @@
+apiVersion: sagemaker.services.k8s.aws/v1alpha1
+kind: Endpoint
+metadata:
+  creationTimestamp: null
```
Comment (surajkota) [inline pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_create_attempted.yaml:4]: metadata.name missing

File: pkg/resource/training_job/testdata/training_job/cr/traij_cmd_invalid_before_create.yaml [hunk 4]
```diff
@@ -0,0 +1,68 @@
+apiVersion: sagemaker.services.k8s.aws/v1alpha1
+kind: FeatureGroup
```
Comment (surajkota) [inline pkg/resource/training_job/testdata/training_job/cr/traij_cmd_invalid_before_create.yaml:2]: mismatch Kind, again how are tests passing?

also, file name inconsistent. problem with short forms but ok. call it trainj

File: pkg/resource/transform_job/custom_delta.go [hunk 5]
```diff
@@ -23,12 +23,16 @@ func customSetDefaults(
 	b *resource,
 ) {
 	// TransformInput is a required field.
-	if ackcompare.IsNil(a.ko.Spec.TransformInput.CompressionType) && ackcompare.IsNotNil(b.ko.Spec.TransformInput.CompressionType) {
-		a.ko.Spec.TransformInput.CompressionType = b.ko.Spec.TransformInput.CompressionType
+	if ackcompare.IsNotNil(a.ko.Spec.TransformInput) && ackcompare.IsNotNil(b.ko.Spec.TransformInput) {
```
Comment (surajkota) [inline pkg/resource/transform_job/custom_delta.go:26]: why this change? TransformInput is a required field and will never be nil

**Review by surajkota — CHANGES_REQUESTED** [review]

File: pkg/resource/endpoint/manager_test_suite_test.go [hunk 6]
```diff
@@ -0,0 +1,131 @@
...
+		awsRegion:    "",
+		sess:         nil,
+		sdkapi:       mockSageMakerAPI,
+		log:          fakeLogger,
+		metrics:      ackmetrics.NewMetrics("sagemaker"),
+	}
+}
+
+// TestEndpointTestSuite runs the test suite for endpoint
+func TestEndpointTestSuite(t *testing.T) {
```
Comment (mbaijal) [inline pkg/resource/endpoint/manager_test_suite_test.go:56]: Is there a way to share this file across resources if we provide the whole path to the test_suite.yaml file ?

File: pkg/resource/endpoint/testdata/test_suite.yaml [hunk 7]
```diff
@@ -0,0 +1,17 @@
...
+     - name: "Create=InvalidInput"
+       description: "Given one of the parameters is invalid, ko.Status shows a terminal condition"
+       given:
+         desired_state: "endpoint/cr/e_cmd_invalid_before_create.yaml"
+         svc_api:
+           - operation: CreateEndpointWithContext
+             error:
+               code: InvalidParameterValue
+               message: "The endpoint name must start with an alphanumeric character."
+       invoke: Create
```
Comment (mbaijal) [inline pkg/resource/endpoint/testdata/test_suite.yaml:14]: Have you tried testing delete/update for any resource ? What is the plan for those tests ?

File: pkg/resource/feature_group/testdata/feature_group/cr/fg_cmd_invalid_before_create.yaml [hunk 8]
```diff
@@ -0,0 +1,17 @@
+apiVersion: sagemaker.services.k8s.aws/v1alpha1
```
Comment (mbaijal) [inline pkg/resource/feature_group/testdata/feature_group/cr/fg_cmd_invalid_before_create.yaml:1]: would it be possible to (eventually) write scripts to generate these files during the unit test runs based on the latest SDK ?

File: pkg/resource/training_job/testdata/training_job/cr/traij_cmd_invalid_before_create.yaml (hunk 4 above)
Comment (mbaijal) [reply pkg/resource/training_job/testdata/training_job/cr/traij_cmd_invalid_before_create.yaml:2]: ++ it is concerning that tests are passing

**Review by mbaijal — APPROVED** [review]
Overall looks ok but there is a lot of manual code (yamls) which is hard to review so do make sure you look into why some testfiles with typos are letting tests pass. 

Good Job !

File: pkg/common/custom_utilities.go [hunk 9]
```diff
@@ -27,3 +38,71 @@ func IsModifyingStatus(
 	}
 	return false
 }
+
+// Checks to see if the contents of a given yaml file, with name stored
+// in expectation, matches the given actualYamlByteArray.
+func IsYamlEqual(expectation *string, actualYamlByteArray *[]byte) bool {
```
[the remark this answers is not in the document]
Comment (BriannaRoskind) [reply pkg/common/custom_utilities.go]: I moved this code to the pkg/testutil/util.go file.

File: pkg/resource/endpoint_config/manager_test_suite_test.go [hunk 10]
```diff
@@ -0,0 +1,126 @@
...
+	"github.com/google/go-cmp/cmp"
+	"github.com/google/go-cmp/cmp/cmpopts"
+	"path/filepath"
+	"testing"
+	ctrlrtzap "sigs.k8s.io/controller-runtime/pkg/log/zap"
+	ackmetrics "github.com/aws-controllers-k8s/runtime/pkg/metrics"
+	"go.uber.org/zap/zapcore"
+)
+
+var	DefaultTimestamp = "0001-01-01T00:00:00Z"
```
[the remark this answers is not in the document]
Comment (BriannaRoskind) [reply pkg/resource/endpoint_config/manager_test_suite_test.go]: These variables were accidentally left in from when I factored out common code. They have now been removed from the individual manager_test_suite_test.go files and reside in the pkg/testutil/util.go file.

File: pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_create_attempted.yaml (hunk 3 above)
Comment (BriannaRoskind) [reply pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_create_attempted.yaml]: The name did not show up because when using yaml Marshal of the input file it found a typo field instead of name, and therefore ignored it.

Since the typo in the input file has been fixed, the output file has now been fixed as well.

File: pkg/resource/transform_job/custom_delta.go (hunk 5 above)
Comment (BriannaRoskind) [reply pkg/resource/transform_job/custom_delta.go:26]: Unit test was breaking due to nil pointer exception. We needed to check that the TransformInput.CompressionType and TransformInput.SplitType were not null before trying to access their fields. (Otherwise we got a nil pointer exception)

File: pkg/resource/endpoint/testdata/test_suite.yaml (hunk 7 above)
Comment (BriannaRoskind) [reply pkg/resource/endpoint/testdata/test_suite.yaml:14]: The delete and readone tests are further down. I have not attempted an update test, though the format should be similar to a create or readone test.

File: pkg/resource/feature_group/testdata/feature_group/cr/fg_cmd_invalid_before_create.yaml (hunk 8 above)
Comment (BriannaRoskind) [reply pkg/resource/feature_group/testdata/feature_group/cr/fg_cmd_invalid_before_create.yaml]: I think it might be possible to write a script to generate a single lifecycle of a resource, similar to what exists in PR #68 (The Feature Group state transition unit tests), based off a base file (which for the feature group state transitions, the base file was the integration test resource feature group yaml file). I essentially did the process by hand, but the unit testing state transition process could probably be replicated in code.

If you then ran the script with multiple base files, you could get many state transition variations depending on the base file, and maybe some optional input about how the sagemaker json mocks should respond. Look at Option number 2 in the sagemaker-controller unit testing design doc to get some other ideas as well on how to allow unit testing without a programmer needing to make and keep track of as many files.

File: pkg/resource/training_job/testdata/training_job/cr/traij_cmd_invalid_before_create.yaml (hunk 4 above)
Comment (BriannaRoskind) [reply pkg/resource/training_job/testdata/training_job/cr/traij_cmd_invalid_before_create.yaml]: It is possible we should validate the desired_state / input yaml file by using the testutil/util.go YamlEqual function.
Suggestion is to do this in a separate PR.

File: pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_before_create.yaml (hunk 2 above)
Comment (BriannaRoskind) [reply pkg/resource/endpoint/testdata/endpoint/cr/e_cmd_invalid_before_create.yaml]: When using yaml Marshal on the input file, it finds a typo field instead of name, and therefore ignored it, and ended up not generating an error.
This could be fixed by validating input files (possibly with YamlEqual).

File: pkg/resource/endpoint/testdata/test_suite.yaml (hunk 7 above)
Comment (mbaijal) [reply pkg/resource/endpoint/testdata/test_suite.yaml:14]: Saw the delete test later yes, thanks!

File: pkg/resource/transform_job/custom_delta.go (hunk 5 above)
Comment (surajkota) [reply pkg/resource/transform_job/custom_delta.go:26]: I don't understand how. this code path is not executed for the test in this PR. there is only one test that is calling Create which is also invalid. Delta is only constructed before doing an update. 

Can you double-check this?

File: pkg/resource/training_job/testdata/training_job/v1alpha1/traij_invalid_create_attempted.yaml [hunk 11]
```diff
@@ -0,0 +1,59 @@
+apiVersion: sagemaker.services.k8s.aws/v1alpha1
```
Comment (surajkota) [inline pkg/resource/training_job/testdata/training_job/v1alpha1/traij_invalid_create_attempted.yaml:1]: use `trainj` and `transformj` in file names instead of traij and tranj

**Review by surajkota — CHANGES_REQUESTED** [review]

**Review by surajkota — APPROVED** [review]
/approve

**Review by mbaijal — APPROVED** [review]
